### DISCUSSION TITLE
Accessing previous `data` during next `data()` hook

### DISCUSSION THREAD

**Discussion by shahyar** [vikejs/vike#1822] [Discussion] [answered]
Hello, great work on the recent updates! One issue I'm running into is in regards to the `data()` hook. Some background:

1. I use `guard()` to protect pages, so I need to have my `pageContext` already loaded with the `user`; this is at `pageContext.data.user`. 
2. I am trying to avoid pageContext.json requests, so I have made `+data.shared.ts` isomorphic.

What happens now is that the initial SSR load correctly runs `data()` and passes that onto the client. However, upon the subsequent CSR navigation, `data()` runs again, but I do not need it to fetch the user again, as I already know they are logged in.

For the rest of my app, my workaround is to store everything in my own ContextProvider, effectively wrapping the useData hook. However, this doesn't solve the problem that `guard()` can only fetch that data from `pageContext` directly, and not from my own data context.

Is there a way to access the last state of `data`/`pageContext` during the `data()` hook function (to return the old data)?

**Comment by brillout** [comment]
I presume you've seen https://vike.dev/pageContext.json#avoid-pagecontext-json-requests?

The `vike/client/client-routing-runtime/renderPageClientSide.ts` file holds `previousPageContext` which we could make available as `pageContext._previousPageContext`. Would that work?

**Reply by brillout** [reply, marked as answer]
Also, in case you didn't see it already: https://vike.dev/guard#execution-order.

**Reply by brillout** [reply]
If I understand your issue correctly, I think a solution would be to extend `+passToClient` to be `(string | { navigation: boolean, value: string })[]` where `navigation: boolean` denotes whether the property is passed to the client upon client-side navigation. I created a feature request for it: https://github.com/vikejs/vike/issues/1842.

**Reply by shahyar** [reply]
> I presume you've seen

Yep, I no longer have pageContext.json. I do something like `async data(pageContext) => !IS_SERVER ? undefined : ...` with `<CustomDataContextProvider data={context.data}>...` which checks if `data` changed to `undefined`, in which case it is ignored.


> `previousPageContext` which we could make available

I think that could do the trick. It would allow `async data(pageContext) => IS_SERVER ? await fetchSessionData() : pageContext._previousPageContext.data`. Right? If it's the exact same object, it won't trigger React update side effects.


> a solution would be to extend `+passToClient`

I'm not using `passToClient` anymore. Although I am migrating _back_ to (`vike-`)`react-query`, and have not yet tested if this causes `pageContext.json` on CSR. Assuming it does not, then that + `vike-react-zustand` will give me pretty much everything I dreamed of as a little boy. 👦🏽

**Reply by shahyar** [reply]
> Also, in case you didn't see it already: https://vike.dev/guard#execution-order.

Ohhhh damn. I missed that `guard` comes before `data` in SSR. Thanks!

~~Is it allowed to modify `pageContext.data` inside `guard`? I am wondering how I'll make the session request and not have to fetch it again in `data`. As this is not within the React tree, I don't have access to my contexts.~~ Edit: I see now. I need to put +data.shared on each protected route instead, and throw directly from there.

**Reply by brillout** [reply]
> `previousPageContext` which we could make available

[Done](https://github.com/vikejs/vike/commit/91f6e68bc2ba90104287e221fa35a796d4ae968b) and pre-released as `0.4.193-commit-91f6e68`.

> I think that could do the trick. It would allow `async data(pageContext) => IS_SERVER ? await fetchSessionData() : pageContext._previousPageContext.data`. Right? If it's the exact same object, it won't trigger React update side effects.

Yes, it should be the same object (`===`). If it isn't then it's a Vike bug (it's actually important that it's the same object).

> I'm not using `passToClient` anymore. Although I am migrating _back_ to (`vike-`)`react-query`, and have not yet tested if this causes `pageContext.json` on CSR. Assuming it does not, then that + `vike-react-zustand` will give me pretty much everything I dreamed of as a little boy. 👦🏽

I'm glad it works for you. Thank you for sponsoring as an individual and, in case your company is up for it, we're looking for company sponsors :green_heart:
